samza-api/src/main/java/org/apache/samza/operators/functions/StreamTableJoinFunction.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.operators.functions;
+
+import org.apache.samza.annotation.InterfaceStability;
+
+
+/**
+ * Joins incoming messages with records from a table by key.
+ *
+ * @param <K>  type of the join key
+ * @param <M>  type of the input message
+ * @param <R>  type of records in the table
+ * @param <OM> type of join results
+ */
+@InterfaceStability.Unstable
+public interface StreamTableJoinFunction<K, M, R, OM> extends InitableFunction, ClosableFunction {
+
+  /**
+   * Joins the provided messages and table record, returns the joined message.
+   *
+   * @param message  the input message
+   * @param record  the table record
+   * @return  the join result
+   */
+  OM apply(M message, R record);
+
+  /**
+   * Get the join key from messages in the input stream.
+   *
+   * @param message  the message in the input stream
+   * @return  the join key
+   */
+  K getFirstKey(M message);
```

Comment:
One other disadvantage of this is that we can't use a simple lambda expression to specify the behavior of a stream table join.

Reply:
@nickpan47 Prateek just mentioned that notifications aren't going out for this review for some reason, so replying late to your question...

There are two options:

1. Move the message -> key function into a separate object that is passed as a parameter to join. This is probably the easiest change that fits with the existing API. Benefits: you can apply the same join function more generally (no implicit assumption about K and you can use potentially wider types for V), and you can use a lambda expression to define the join function.

2. I really like the Beam approach, but it is a significant departure from the current API design. You can get a sense for it with their [GroupByKey](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/GroupByKey.java#L121) operator. The way they preserve type safety is that they put the constraints on the operator and use apply on the stream instead of putting the operator directly on the stream. For example, GroupByKey is defined as a transform of `PCollection<KV<K, V>>` to `PCollection<KV<K, Iterable<V>>>`, where `PCollection` is similar to `MessageStream`.

Reply:
@cpettitt-linkedin thanks! Wei and I had a discussion offline as well and we also think that one way to deal w/ that is to apply types before the join. Following the same approach in your second point, I think that one way to handle this is to define the input types in JoinFunction as KV pairs and when we use JoinFunction in MessageStream.join(), we can infer that the input MessageStream type should be KV as well. That's a very similar model as Beam's apply of GroupByKey method: i.e. JoinFunction<KV<K, ?>, KV<K,?>, ?> ==> GroupByKey and MessageStream<KV<K, ?>>.join() ==> PCollection<KV<K, ?>>.apply().

@weisong44 can we implement stream-to-table join using the suggested pattern? We should definitely open another JIRA to change the stream-to-stream join to the same way separately.

Reply:
Yes, this makes sense. Using KV<K, V> makes the API more clear and concise. Changes made.